File: tests/test_call_conv.py
#!/usr/bin/env -S uv run
"""
Test: test_call_conv
Tests calling convention enum values and instruction call conv get/set
via Python bindings.

This is the Python equivalent of tests/test_call_conv.cpp.
Output should match the C++ golden master test.

LLVM APIs covered (via Python bindings):
- CallConv enum (all values including new: GHC, HiPE, etc.)
- instruction_call_conv property (get/set)
- calling_conv property on functions
"""

import sys
import llvm


def call_conv_name(cc: llvm.CallConv) -> str:
    mapping = {
        llvm.CallConv.C: "ccc",
        llvm.CallConv.Fast: "fastcc",
        llvm.CallConv.Cold: "coldcc",
        llvm.CallConv.X86Stdcall: "x86_stdcallcc",
        llvm.CallConv.X86Fastcall: "x86_fastcallcc",
        llvm.CallConv.GHC: "ghccc",
        llvm.CallConv.HiPE: "cc11",
        llvm.CallConv.PreserveMost: "preserve_mostcc",
        llvm.CallConv.PreserveAll: "preserve_allcc",
        llvm.CallConv.Swift: "swiftcc",
        llvm.CallConv.CXX_FAST_TLS: "cxx_fast_tlscc",
        llvm.CallConv.X86ThisCall: "x86_thiscallcc",
        llvm.CallConv.X86_64_SysV: "x86_64_sysvcc",
        llvm.CallConv.Win64: "win64cc",
        llvm.CallConv.X86VectorCall: "x86_vectorcallcc",
        llvm.CallConv.X86RegCall: "x86_regcallcc",
    }
    return mapping.get(cc, "unknown")


def main():
    with llvm.create_context() as ctx:
        with ctx.create_module("test_call_conv") as mod:
            i32 = ctx.types.i32
            void_ty = ctx.types.void

            # ==========================================
            # Test 1: CallConv enum values
            # ==========================================
            convs = [
                (llvm.CallConv.C, "C"),
                (llvm.CallConv.Fast, "Fast"),
                (llvm.CallConv.Cold, "Cold"),
                (llvm.CallConv.X86Stdcall, "X86Stdcall"),
                (llvm.CallConv.X86Fastcall, "X86Fastcall"),
                (llvm.CallConv.GHC, "GHC"),
                (llvm.CallConv.HiPE, "HiPE"),
                (llvm.CallConv.PreserveMost, "PreserveMost"),
                (llvm.CallConv.PreserveAll, "PreserveAll"),
                (llvm.CallConv.Swift, "Swift"),
                (llvm.CallConv.CXX_FAST_TLS, "CXX_FAST_TLS"),
                (llvm.CallConv.X86ThisCall, "X86ThisCall"),
                (llvm.CallConv.X86_64_SysV, "X86_64_SysV"),
                (llvm.CallConv.Win64, "Win64"),
                (llvm.CallConv.X86VectorCall, "X86VectorCall"),
                (llvm.CallConv.X86RegCall, "X86RegCall"),
            ]

            # ==========================================
            # Test 2: Declare a callee and callers with different CCs
            # ==========================================

            # Callee: i32 callee(i32)
            callee_ty = ctx.types.function(i32, [i32])
            callee = mod.add_function("callee", callee_ty)

            # Caller: i32 caller_default(i32) - default CC call
            caller_default = mod.add_function("caller_default", callee_ty)
            cd_param = caller_default.get_param(0)
            cd_param.name = "x"
            cd_entry = caller_default.append_basic_block("entry")

            with cd_entry.create_builder() as builder:
                cd_call = builder.call(callee_ty, callee, [cd_param], "result")
                builder.ret(cd_call)

                # Read default call conv
                default_cc = cd_call.instruction_call_conv

                # Caller: i32 caller_fastcc(i32) - fastcc call
                caller_fastcc = mod.add_function("caller_fastcc", callee_ty)
                cf_param = caller_fastcc.get_param(0)
                cf_param.name = "x"
                cf_entry = caller_fastcc.append_basic_block("entry")
                builder.position_at_end(cf_entry)
                cf_call = builder.call(callee_ty, callee, [cf_param], "result")
                cf_call.instruction_call_conv = llvm.CallConv.Fast
                builder.ret(cf_call)
                fast_cc = cf_call.instruction_call_conv

                # Caller: i32 caller_coldcc(i32) - coldcc call
                caller_coldcc = mod.add_function("caller_coldcc", callee_ty)
                cc_param = caller_coldcc.get_param(0)
                cc_param.name = "x"
                cc_entry = caller_coldcc.append_basic_block("entry")
                builder.position_at_end(cc_entry)
                cc_call = builder.call(callee_ty, callee, [cc_param], "result")
                cc_call.instruction_call_conv = llvm.CallConv.Cold
                builder.ret(cc_call)
                cold_cc = cc_call.instruction_call_conv

                # Caller: i32 caller_ghccc(i32) - ghccc call
                caller_ghccc = mod.add_function("caller_ghccc", callee_ty)
                cg_param = caller_ghccc.get_param(0)
                cg_param.name = "x"
                cg_entry = caller_ghccc.append_basic_block("entry")
                builder.position_at_end(cg_entry)
                cg_call = builder.call(callee_ty, callee, [cg_param], "result")
                cg_call.instruction_call_conv = llvm.CallConv.GHC
                builder.ret(cg_call)
                ghc_cc = cg_call.instruction_call_conv

                # Caller: i32 caller_swiftcc(i32) - swiftcc call
                caller_swiftcc = mod.add_function("caller_swiftcc", callee_ty)
                cs_param = caller_swiftcc.get_param(0)
                cs_param.name = "x"
                cs_entry = caller_swiftcc.append_basic_block("entry")
                builder.position_at_end(cs_entry)
                cs_call = builder.call(callee_ty, callee, [cs_param], "result")
                cs_call.instruction_call_conv = llvm.CallConv.Swift
                builder.ret(cs_call)
                swift_cc = cs_call.instruction_call_conv

            # Verify module
            if not mod.verify():
                print(
                    f"; Verification failed: {mod.get_verification_error()}",
                    file=sys.stderr,
                )
                return 1

            # Print diagnostic comments
            print("; Test: test_call_conv")
            print(";")

            print("; CallConv enum values:")
            for cc_val, name in convs:
                print(f";   {name} = {cc_val.value} ({call_conv_name(cc_val)})")

            print(";")
            print("; Instruction call conv tests:")
            print(
                f";   default call conv: {default_cc.value} ({call_conv_name(default_cc)})"
            )
            print(
                f";   after set fastcc: {fast_cc.value} ({call_conv_name(fast_cc)})"
            )
            print(
                f";   after set coldcc: {cold_cc.value} ({call_conv_name(cold_cc)})"
            )
            print(
                f";   after set ghccc: {ghc_cc.value} ({call_conv_name(ghc_cc)})"
            )
            print(
                f";   after set swiftcc: {swift_cc.value} ({call_conv_name(swift_cc)})"
            )

            print()

            # Print module IR
            print(mod.to_string(), end="")

    return 0


if __name__ == "__main__":
    sys.exit(main())

File: tests/test_call_conv.cpp
/**
 * Test: test_call_conv
 * Tests calling convention enum values and instruction call conv get/set
 *
 * LLVM-C APIs covered:
 * - LLVMGetInstructionCallConv()
 * - LLVMSetInstructionCallConv()
 * - LLVMSetFunctionCallConv(), LLVMGetFunctionCallConv()
 * - Various LLVMCallConv enum values
 */

#include <cstdio>
#include <llvm-c/Analysis.h>
#include <llvm-c/Core.h>

const char *call_conv_name(unsigned cc) {
  switch (cc) {
  case LLVMCCallConv: return "ccc";
  case LLVMFastCallConv: return "fastcc";
  case LLVMColdCallConv: return "coldcc";
  case LLVMX86StdcallCallConv: return "x86_stdcallcc";
  case LLVMX86FastcallCallConv: return "x86_fastcallcc";
  case LLVMGHCCallConv: return "ghccc";
  case LLVMHiPECallConv: return "cc11";
  case LLVMPreserveMostCallConv: return "preserve_mostcc";
  case LLVMPreserveAllCallConv: return "preserve_allcc";
  case LLVMSwiftCallConv: return "swiftcc";
  case LLVMCXXFASTTLSCallConv: return "cxx_fast_tlscc";
  case LLVMX86ThisCallCallConv: return "x86_thiscallcc";
  case LLVMX8664SysVCallConv: return "x86_64_sysvcc";
  case LLVMWin64CallConv: return "win64cc";
  case LLVMX86VectorCallCallConv: return "x86_vectorcallcc";
  case LLVMX86RegCallCallConv: return "x86_regcallcc";
  default: return "unknown";
  }
}

int main() {
  LLVMContextRef ctx = LLVMContextCreate();
  LLVMModuleRef mod =
      LLVMModuleCreateWithNameInContext("test_call_conv", ctx);

  LLVMTypeRef i32 = LLVMInt32TypeInContext(ctx);
  LLVMTypeRef void_ty = LLVMVoidTypeInContext(ctx);

  LLVMBuilderRef builder = LLVMCreateBuilderInContext(ctx);

  // ==========================================
  // Test 1: CallConv enum values
  // ==========================================
  struct {
    unsigned value;
    const char *name;
  } convs[] = {
      {LLVMCCallConv, "C"},
      {LLVMFastCallConv, "Fast"},
      {LLVMColdCallConv, "Cold"},
      {LLVMX86StdcallCallConv, "X86Stdcall"},
      {LLVMX86FastcallCallConv, "X86Fastcall"},
      {LLVMGHCCallConv, "GHC"},
      {LLVMHiPECallConv, "HiPE"},
      {LLVMPreserveMostCallConv, "PreserveMost"},
      {LLVMPreserveAllCallConv, "PreserveAll"},
      {LLVMSwiftCallConv, "Swift"},
      {LLVMCXXFASTTLSCallConv, "CXX_FAST_TLS"},
      {LLVMX86ThisCallCallConv, "X86ThisCall"},
      {LLVMX8664SysVCallConv, "X86_64_SysV"},
      {LLVMWin64CallConv, "Win64"},
      {LLVMX86VectorCallCallConv, "X86VectorCall"},
      {LLVMX86RegCallCallConv, "X86RegCall"},
  };

  // ==========================================
  // Test 2: Declare a callee and a caller that uses call with different CCs
  // ==========================================

  // Callee: i32 callee(i32)
  LLVMTypeRef callee_params[] = {i32};
  LLVMTypeRef callee_ty = LLVMFunctionType(i32, callee_params, 1, 0);
  LLVMValueRef callee = LLVMAddFunction(mod, "callee", callee_ty);

  // Caller: i32 caller_default(i32) - default CC call
  LLVMValueRef caller_default =
      LLVMAddFunction(mod, "caller_default", callee_ty);
  LLVMValueRef cd_param = LLVMGetParam(caller_default, 0);
  LLVMSetValueName2(cd_param, "x", 1);
  LLVMBasicBlockRef cd_entry =
      LLVMAppendBasicBlockInContext(ctx, caller_default, "entry");
  LLVMPositionBuilderAtEnd(builder, cd_entry);
  LLVMValueRef cd_args[] = {cd_param};
  LLVMValueRef cd_call =
      LLVMBuildCall2(builder, callee_ty, callee, cd_args, 1, "result");
  LLVMBuildRet(builder, cd_call);

  // Read default call conv
  unsigned default_cc = LLVMGetInstructionCallConv(cd_call);

  // Caller: i32 caller_fastcc(i32) - fastcc call
  LLVMValueRef caller_fastcc =
      LLVMAddFunction(mod, "caller_fastcc", callee_ty);
  LLVMValueRef cf_param = LLVMGetParam(caller_fastcc, 0);
  LLVMSetValueName2(cf_param, "x", 1);
  LLVMBasicBlockRef cf_entry =
      LLVMAppendBasicBlockInContext(ctx, caller_fastcc, "entry");
  LLVMPositionBuilderAtEnd(builder, cf_entry);
  LLVMValueRef cf_args[] = {cf_param};
  LLVMValueRef cf_call =
      LLVMBuildCall2(builder, callee_ty, callee, cf_args, 1, "result");
  LLVMSetInstructionCallConv(cf_call, LLVMFastCallConv);
  LLVMBuildRet(builder, cf_call);

  unsigned fast_cc = LLVMGetInstructionCallConv(cf_call);

  // Caller: i32 caller_coldcc(i32) - coldcc call
  LLVMValueRef caller_coldcc =
      LLVMAddFunction(mod, "caller_coldcc", callee_ty);
  LLVMValueRef cc_param = LLVMGetParam(caller_coldcc, 0);
  LLVMSetValueName2(cc_param, "x", 1);
  LLVMBasicBlockRef cc_entry =
      LLVMAppendBasicBlockInContext(ctx, caller_coldcc, "entry");
  LLVMPositionBuilderAtEnd(builder, cc_entry);
  LLVMValueRef cc_args[] = {cc_param};
  LLVMValueRef cc_call =
      LLVMBuildCall2(builder, callee_ty, callee, cc_args, 1, "result");
  LLVMSetInstructionCallConv(cc_call, LLVMColdCallConv);
  LLVMBuildRet(builder, cc_call);

  unsigned cold_cc = LLVMGetInstructionCallConv(cc_call);

  // Caller: i32 caller_ghccc(i32) - ghccc call
  LLVMValueRef caller_ghccc =
      LLVMAddFunction(mod, "caller_ghccc", callee_ty);
  LLVMValueRef cg_param = LLVMGetParam(caller_ghccc, 0);
  LLVMSetValueName2(cg_param, "x", 1);
  LLVMBasicBlockRef cg_entry =
      LLVMAppendBasicBlockInContext(ctx, caller_ghccc, "entry");
  LLVMPositionBuilderAtEnd(builder, cg_entry);
  LLVMValueRef cg_args[] = {cg_param};
  LLVMValueRef cg_call =
      LLVMBuildCall2(builder, callee_ty, callee, cg_args, 1, "result");
  LLVMSetInstructionCallConv(cg_call, LLVMGHCCallConv);
  LLVMBuildRet(builder, cg_call);

  unsigned ghc_cc = LLVMGetInstructionCallConv(cg_call);

  // Caller: i32 caller_swiftcc(i32) - swiftcc call
  LLVMValueRef caller_swiftcc =
      LLVMAddFunction(mod, "caller_swiftcc", callee_ty);
  LLVMValueRef cs_param = LLVMGetParam(caller_swiftcc, 0);
  LLVMSetValueName2(cs_param, "x", 1);
  LLVMBasicBlockRef cs_entry =
      LLVMAppendBasicBlockInContext(ctx, caller_swiftcc, "entry");
  LLVMPositionBuilderAtEnd(builder, cs_entry);
  LLVMValueRef cs_args[] = {cs_param};
  LLVMValueRef cs_call =
      LLVMBuildCall2(builder, callee_ty, callee, cs_args, 1, "result");
  LLVMSetInstructionCallConv(cs_call, LLVMSwiftCallConv);
  LLVMBuildRet(builder, cs_call);

  unsigned swift_cc = LLVMGetInstructionCallConv(cs_call);

  LLVMDisposeBuilder(builder);

  // Verify module
  char *error = nullptr;
  if (LLVMVerifyModule(mod, LLVMReturnStatusAction, &error)) {
    fprintf(stderr, "; Verification failed: %s\n", error);
    LLVMDisposeMessage(error);
    LLVMDisposeModule(mod);
    LLVMContextDispose(ctx);
    return 1;
  }
  LLVMDisposeMessage(error);

  // Print diagnostic comments
  printf("; Test: test_call_conv\n");
  printf(";\n");

  printf("; CallConv enum values:\n");
  for (size_t i = 0; i < sizeof(convs) / sizeof(convs[0]); i++) {
    printf(";   %s = %u (%s)\n", convs[i].name, convs[i].value,
           call_conv_name(convs[i].value));
  }

  printf(";\n");
  printf("; Instruction call conv tests:\n");
  printf(";   default call conv: %u (%s)\n", default_cc,
         call_conv_name(default_cc));
  printf(";   after set fastcc: %u (%s)\n", fast_cc,
         call_conv_name(fast_cc));
  printf(";   after set coldcc: %u (%s)\n", cold_cc,
         call_conv_name(cold_cc));
  printf(";   after set ghccc: %u (%s)\n", ghc_cc,
         call_conv_name(ghc_cc));
  printf(";   after set swiftcc: %u (%s)\n", swift_cc,
         call_conv_name(swift_cc));

  printf("\n");

  // Print module IR
  char *ir = LLVMPrintModuleToString(mod);
  printf("%s", ir);
  LLVMDisposeMessage(ir);

  // Cleanup
  LLVMDisposeModule(mod);
  LLVMContextDispose(ctx);

  return 0;
}
